### DISCUSSION TITLE
Traffic

### DISCUSSION THREAD

**Discussion by itrack** [valhalla/valhalla#4412] [Q&A]
Hi!

The traffic system seems poorly documented and hard to understand how it works. (at least it was hard for me 😅)
I'm not very familiar with Valhalla in detail, I'm more of a docker run kind of user.

I opened this topic to ask if you can explain me how I can integrate a traffic system into Valhalla. From what I understand so far, this is done using a csv, can you give me an example of a csv?

Can the integration be done during a certain period of the day/week, i.e. can the traffic be predictive?

Thank you very much!

**Comment by nilsnolde** [comment]
True, it's not very well documented in the repo. And documentation is hard, also this kind of documentation (on my side) has the lowest priority. Basically, IMO if a company can shell out the cash to acquire this kind of data, they can also pay for implementation/consultation from people who went through that big-ish effort themselves.

Some refs:
- https://github.com/valhalla/valhalla/pull/4259
- smth I wrote a while back: https://gis-ops.com/traffic-in-valhalla/ (though no guarantee there, that's a long time ago)
- there's also some repo showing it bit more which you should be able to find (can't remember the name)
- look at the tests, historical/predictive traffic is covered and shows much of its implementation

And the inevitable pitch: enquiry@gis-ops.com :smile:

**Reply by matthiscock** [reply]
We want to do just that, and have emailed in, but not heard back!

**Comment by itrack** [comment]
Thank you! 
Being more of a PHP programmer than a C++ programmer it is a bit hard to understand but I will try to see the tests and understand how it works :)
If there are others in the community who would be willing to help on this topic it would be very helpful.
And I'll publish here if they make any discoveries 😃
